File: python/polars_statistics/exprs/nonparametric.py
"""Non-parametric statistical tests as Polars expressions."""

from __future__ import annotations

from pathlib import Path
from typing import Union

import polars as pl
from polars.plugins import register_plugin_function

LIB = Path(__file__).parent.parent


def mann_whitney_u(
    x: Union[pl.Expr, str],
    y: Union[pl.Expr, str],
) -> pl.Expr:
    """
    Perform Mann-Whitney U test (Wilcoxon rank-sum test).

    A non-parametric test for testing whether two independent samples
    were drawn from the same distribution. Uses two-sided test.

    Parameters
    ----------
    x : pl.Expr or str
        First sample expression or column name.
    y : pl.Expr or str
        Second sample expression or column name.

    Returns
    -------
    pl.Expr
        Expression returning struct{statistic: f64, p_value: f64}

    Examples
    --------
    >>> import polars as pl
    >>> import polars_statistics as ps
    >>>
    >>> df = pl.DataFrame({
    ...     "x": [1.0, 2.0, 3.0, 4.0, 5.0],
    ...     "y": [2.0, 4.0, 6.0, 8.0, 10.0],
    ... })
    >>>
    >>> df.select(ps.mann_whitney_u("x", "y"))
    """
    if isinstance(x, str):
        x = pl.col(x)
    if isinstance(y, str):
        y = pl.col(y)

    x_clean = x.filter(x.is_finite())
    y_clean = y.filter(y.is_finite())

    return register_plugin_function(
        plugin_path=LIB,
        function_name="pl_mann_whitney_u",
        args=[x_clean, y_clean],
        returns_scalar=True,
    )


def wilcoxon_signed_rank(
    x: Union[pl.Expr, str],
    y: Union[pl.Expr, str],
) -> pl.Expr:
    """
    Perform Wilcoxon signed-rank test.

    A non-parametric test for paired samples, testing whether
    paired differences come from a symmetric distribution around zero.
    Uses two-sided test.

    Parameters
    ----------
    x : pl.Expr or str
        First sample expression or column name.
    y : pl.Expr or str
        Second sample expression or column name.

    Returns
    -------
    pl.Expr
        Expression returning struct{statistic: f64, p_value: f64}
    """
    if isinstance(x, str):
        x = pl.col(x)
    if isinstance(y, str):
        y = pl.col(y)

    x_clean = x.filter(x.is_finite() & y.is_finite())
    y_clean = y.filter(x.is_finite() & y.is_finite())

    return register_plugin_function(
        plugin_path=LIB,
        function_name="pl_wilcoxon_signed_rank",
        args=[x_clean, y_clean],
        returns_scalar=True,
    )


def kruskal_wallis(
    *groups: Union[pl.Expr, str],
) -> pl.Expr:
    """
    Perform Kruskal-Wallis H test.

    A non-parametric test for testing whether samples from two or more
    groups originate from the same distribution.

    Parameters
    ----------
    *groups : pl.Expr or str
        Two or more sample expressions or column names.

    Returns
    -------
    pl.Expr
        Expression returning struct{statistic: f64, p_value: f64}

    Examples
    --------
    >>> import polars as pl
    >>> import polars_statistics as ps
    >>>
    >>> df = pl.DataFrame({
    ...     "group1": [1.0, 2.0, 3.0],
    ...     "group2": [4.0, 5.0, 6.0],
    ...     "group3": [7.0, 8.0, 9.0],
    ... })
    >>>
    >>> df.select(ps.kruskal_wallis("group1", "group2", "group3"))
    """
    cleaned_groups = []
    for g in groups:
        if isinstance(g, str):
            g = pl.col(g)
        cleaned_groups.append(g.filter(g.is_finite()))

    return register_plugin_function(
        plugin_path=LIB,
        function_name="pl_kruskal_wallis",
        args=cleaned_groups,
        returns_scalar=True,
    )


def brunner_munzel(
    x: Union[pl.Expr, str],
    y: Union[pl.Expr, str],
    alternative: str = "two-sided",
) -> pl.Expr:
    """
    Perform Brunner-Munzel test for stochastic equality.

    A robust non-parametric test for comparing two independent samples
    that doesn't assume equal variances or shape of distributions.
    Tests whether P(X < Y) = 0.5.

    Parameters
    ----------
    x : pl.Expr or str
        First sample expression or column name.
    y : pl.Expr or str
        Second sample expression or column name.
    alternative : {"two-sided", "less", "greater"}, default "two-sided"
        Alternative hypothesis direction.

    Returns
    -------
    pl.Expr
        Expression returning struct{statistic: f64, p_value: f64}

    Examples
    --------
    >>> import polars as pl
    >>> import polars_statistics as ps
    >>>
    >>> df = pl.DataFrame({
    ...     "x": [1.0, 2.0, 3.0, 4.0, 5.0],
    ...     "y": [2.0, 4.0, 6.0, 8.0, 10.0],
    ... })
    >>>
    >>> df.select(ps.brunner_munzel("x", "y"))
    """
    if isinstance(x, str):
        x = pl.col(x)
    if isinstance(y, str):
        y = pl.col(y)

    x_clean = x.filter(x.is_finite())
    y_clean = y.filter(y.is_finite())

    return register_plugin_function(
        plugin_path=LIB,
        function_name="pl_brunner_munzel",
        args=[
            x_clean,
            y_clean,
            pl.lit(alternative, dtype=pl.String),
        ],
        returns_scalar=True,
    )

File: src/expressions/nonparametric.rs
//! Non-parametric statistical test expressions.

use polars::prelude::*;
use pyo3_polars::derive::polars_expr;

use anofox_statistics::{mann_whitney_u, wilcoxon_signed_rank, kruskal_wallis};

use crate::expressions::output_types::{generic_stats_output, stats_output_dtype};

/// Mann-Whitney U test (two-sided)
#[polars_expr(output_type_func=stats_output_dtype)]
fn pl_mann_whitney_u(inputs: &[Series]) -> PolarsResult<Series> {
    let x = inputs[0].f64()?;
    let y = inputs[1].f64()?;
    // Note: alternative parameter is ignored as the underlying library only supports two-sided
    // let _alt_str = inputs[2].str()?.get(0).unwrap_or("two-sided");

    let x_vec: Vec<f64> = x.into_no_null_iter().collect();
    let y_vec: Vec<f64> = y.into_no_null_iter().collect();

    match mann_whitney_u(&x_vec, &y_vec) {
        Ok(result) => generic_stats_output(result.statistic, result.p_value, "mann_whitney_u"),
        Err(_) => generic_stats_output(f64::NAN, f64::NAN, "mann_whitney_u"),
    }
}

/// Wilcoxon signed-rank test (two-sided)
#[polars_expr(output_type_func=stats_output_dtype)]
fn pl_wilcoxon_signed_rank(inputs: &[Series]) -> PolarsResult<Series> {
    let x = inputs[0].f64()?;
    let y = inputs[1].f64()?;
    // Note: alternative parameter is ignored as the underlying library only supports two-sided
    // let _alt_str = inputs[2].str()?.get(0).unwrap_or("two-sided");

    let x_vec: Vec<f64> = x.into_no_null_iter().collect();
    let y_vec: Vec<f64> = y.into_no_null_iter().collect();

    match wilcoxon_signed_rank(&x_vec, &y_vec) {
        Ok(result) => generic_stats_output(result.statistic, result.p_value, "wilcoxon_signed_rank"),
        Err(_) => generic_stats_output(f64::NAN, f64::NAN, "wilcoxon_signed_rank"),
    }
}

/// Kruskal-Wallis H test
#[polars_expr(output_type_func=stats_output_dtype)]
fn pl_kruskal_wallis(inputs: &[Series]) -> PolarsResult<Series> {
    // This takes multiple groups as separate series
    let groups: Vec<Vec<f64>> = inputs
        .iter()
        .map(|s| s.f64().map(|ca| ca.into_no_null_iter().collect()))
        .collect::<PolarsResult<Vec<_>>>()?;

    let group_refs: Vec<&[f64]> = groups.iter().map(|g| g.as_slice()).collect();

    match kruskal_wallis(&group_refs) {
        Ok(result) => generic_stats_output(result.statistic, result.p_value, "kruskal_wallis"),
        Err(_) => generic_stats_output(f64::NAN, f64::NAN, "kruskal_wallis"),
    }
}
